### ISSUE TITLE
[NuGet.org Bug]: 

### ISSUE THREAD

**Issue by drewgillies** [NuGet/NuGetGallery#9791]
### Impact

Other

### Describe the bug

Statistics on the search page don't match statistics on the package details page.

### Repro Steps

Search for Fabulous Scheduler on NuGet.org and view the results: https://www.nuget.org/packages?q=fabulous+scheduler
Click on the FabulousScheduler link on the results page and view the stats on the package details page: https://www.nuget.org/packages/FabulousScheduler

They don't match.

### Expected Behavior

Both screen show the same download count.

### Screenshots

Search results:
![image](https://github.com/NuGet/NuGetGallery/assets/14225979/b4367bef-acd3-48a0-9c9b-d7ff714aeb7e)

Package details:
![image](https://github.com/NuGet/NuGetGallery/assets/14225979/08dccb02-94a8-4f06-963d-cc593b4b31f0)

**Comment by swharden** [comment]
Possibly related, the primary query API has been serving stale download counts for a few weeks, but the secondary endpoint serves higher download counts.

e.g., Compare `totalDownloads` of [ussc](https://azuresearch-ussc.nuget.org/query?take=1&q=scottplot) vs [usnc](https://azuresearch-usnc.nuget.org/query?take=1&q=scottplot). According to my records, the last time the primary endpoint updated download counts was Jan 19, 2024 (22 days ago), but the secondary one seems to be working as expected.

A side effect of this issue is that NuGet trends graphs have leveled-out for the last few weeks. E.g., https://nugettrends.com/packages?months=6&ids=NUnit

![image](https://github.com/NuGet/NuGetGallery/assets/4165489/b3ff7144-edf4-4484-b4c8-ac493d4e5383)

I hope this information is helpful. Thanks NuGet team for all you do! 🚀

**Comment by joelverhagen** [comment]
If you independently check download counts on the search API vs. the gallery, you will often see a different number. This is by design because there is no shared, live cache that gallery and search services depend on (and we don't want the headache of the SPOF). 

But the problem related to rendering in the gallery, for example search results show in gallery vs. the package details page could be resolved by replacing the search API download count with what the gallery knows via its own cache. This would at least make gallery self-consistent.

**Comment by jodydonetti** [comment]
> But the problem related to rendering in the gallery, for example search results show in gallery vs. the package details page could be resolved by replacing the search API download count with what the gallery knows via its own cache. This would at least make gallery self-consistent.

True but, and I haven't looked at the code yet so I may be wrong about this, I smell a potential "SELECT N+1" problem here...

Will look into that as soon as possible.

**Comment by joelverhagen** [comment]
> True but, and I haven't looked at the code yet so I may be wrong about this, I smell at least a potential "SELECT N+1" problem here...

Yes, depending on the implementation of the cache you're totally right. Today the cache is a giant in-memory dictionary where point reads are "free". So doing 20-30 download count reads as a fix-up when rendering the search page could work. But a more traditional read-through would have that problem.

**Comment by jodydonetti** [comment]
Ah, good call! I'm thinking about something right now, will update.
